cpp/arcticdb/vector_db/vector_db.cpp

```diff
@@ -0,0 +1,102 @@
+                                                     Composite<ProcessingSegment> &&p) const {
+        auto procs = std::move(p);
+        TopK top_k(k_);
+        // We expect a vector in each column.
+        // top_k_ is updated as we read more vectors. We want to efficiently remove
+        // the 'worst' (least similar) vector as we go along to ensure we have the
+        // top k, not the top k+1.
+        auto lp = 2;
+        // Let's pretend that only the lp-norms exist. We'll use p=2 for now as a default.
+
+        procs.broadcast(
+                [&store, &top_k, lp, this](const ProcessingSegment &proc) {
+                    for (const auto& slice_and_key: proc.data()) {
+                        std::vector <std::shared_ptr<Column>> columns = slice_and_key.segment(store).columns();
+                        for (auto&& [idx, col]: folly::enumerate(columns)) {
+                            internal::check<ErrorCode::E_ASSERTION_FAILURE>(
+                                    static_cast<long unsigned>(col->row_count()) == this->query_vector_.size(),
+                                    "Expected vector of length {}, got vector of length {}.",
+                                    col->row_count(),
+                                    this->query_vector_.size());
+                            col->type().visit_tag([&top_k, &slice_and_key, &store, &col=col, idx=idx, lp, this](auto type_desc_tag) {
+                                using TypeDescriptorTag = decltype(type_desc_tag);
+                                using RawType = typename TypeDescriptorTag::DataTypeTag::raw_type;
+
+                                if constexpr (is_floating_point_type(TypeDescriptorTag::DataTypeTag::data_type)) {
+
+                                    ColumnData col_data = col->data();
+                                    double sum_differences_exponentiated = 0;
+                                    auto j = 0u;
+
+                                    while (auto block = col_data.next<TypeDescriptorTag>()) {
+                                        auto ptr = reinterpret_cast<const RawType *>(block.value().data());
+                                        for (auto progress_in_block = 0u; progress_in_block <
+                                                                          block.value().row_count(); ++progress_in_block, ++j, ++ptr) {
+                                            sum_differences_exponentiated += pow(
+                                                    abs(*ptr - this->query_vector_[j]), lp);
+                                        }
+                                    }
+
+                                    top_k.try_insert(NamedColumnSortedByDistance(
+                                            sum_differences_exponentiated,
+                                            col,
+                                            slice_and_key.segment(store).field(idx).name()));
```

Comment:
Just noticed this is probably quite inefficient since we check the name of each vector even if we don't insert it, but I suppose I can refactor this when parallelising (just dealing iwth all obvious performance improvements).

Reply:
It's just a very small reference class, nothing gets copied so unless it actually shows up in a performance testing I think it's probably okay